File: searchdata.py
import os

#Reads a file, O(n) if the file does not have a bound
def readFile(path):
    file = open(path,"r")
    content = file.readlines()
    file.close()
    return content

#Returns outgoing links, O(n) as the file gets larger as there are more links
def get_outgoing_links(URL): 
    new_url = URL.replace('/','{')[:-5].replace(':','}')
    if os.path.exists(os.path.join("pages",new_url)):
        return readFile(os.path.join("pages",new_url,"outgoinglinks.txt"))[0].split()
    return None

#Returns incoming links, O(n) as the file gets larger as there are more links
def get_incoming_links(URL):
    new_url = URL.replace('/','{')[:-5].replace(':','}')
    if os.path.exists(os.path.join("pages",new_url)):
        return readFile(os.path.join("pages",new_url,"incominglinks.txt"))[0].split()
    return None

#Returns the pageRank of a page, O(1) since the file doesn't get larger as there are more urls
def get_page_rank(URL):
    new_url = URL.replace('/','{')[:-5].replace(':','}')
    if os.path.exists(os.path.join("pages",new_url)):
        return float(readFile(os.path.join("pages",new_url,"pageRank.txt"))[0])
    return -1

#Returns the title of a page, O(1) since the file doesn't get larger as there are more urls
def get_title(URL):
    new_url = URL.replace('/','{')[:-5].replace(':','}')
    if os.path.exists(os.path.join("pages",new_url)):
        return str(readFile(os.path.join("pages",new_url,"title.txt"))[0])
    return None

#Returns the idf of a page, O(1) since the file doesn't get larger as there are more urls
def get_idf(word):
    if os.path.exists(path:=os.path.join("Idfs",f"{word}.txt")):
        return float(readFile(path)[0])
    return 0

#Returns the tf of a page, O(1) since the file doesn't get larger as there are more urls
def get_tf(URL,word):
    new_url = URL.replace('/','{')[:-5].replace(':','}')
    if os.path.exists(urlPath:=os.path.join("pages",new_url)):
        if os.path.exists(path:=os.path.join(urlPath,"countAll",f"{word}.txt")):
            return float(readFile(path)[0].split()[2])
    return 0

#Returns the tf-idf of a page, O(1) since the file doesn't get larger as there are more urls
def get_tf_idf(URL, word):
    new_url = URL.replace('/','{')[:-5].replace(':','}')
    if os.path.exists(urlPath:=os.path.join("pages",new_url)):
        if os.path.exists(path:=os.path.join(urlPath,"countAll",f"{word}.txt")):
            return float(readFile(path)[0].split()[1])
    return 0

File: CourseProject1406/src/course/project/SearchData.java
package course.project;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class SearchData implements ProjectTester {

	private List<SearchResult> searchResults;
	private Crawler crawler;
	private Search searchObj

	public void initialize(){
		File file = new File("pages");
		if (file.exists()){
			Crawler.deleteFolder(file);
		}
		file.mkdir();

		file = new File("IDFs");
		if (file.exists()){
			Crawler.deleteFolder(file);
		}
		file.mkdir();
	}

	public void crawl(String seedURL){
		crawler = new Crawler();
		crawler.crawl(seedURL);
	}

	public List<String> getOutgoingLinks(String URL){
			String newURL = URL.replace('/','{').replace(':','}').replace('.','(');
			File file = new File("pages" + File.separator + newURL);
			if (file.exists()){
				return List.of(readFile("pages" + File.separator + newURL + File.separator + "outgoinglinks.txt").split(" "));
			}
		return null;
	}

	public List<String> getIncomingLinks(String URL){
		String newURL = URL.replace('/','{').replace(':','}').replace('.','(');
		File file = new File("pages" + File.separator + newURL);
		if (file.exists()){
			return List.of(readFile("pages" + File.separator + newURL + File.separator + "incominglinks.txt").split(" "));
		}
		return null;
	}

	public double getPageRank(String URL){
		String newURL = URL.replace('/','{').replace(':','}').replace('.','(');
		File file = new File("pages" + File.separator + newURL);
		if (file.exists()){
			return Double.parseDouble(readFile("pages" + File.separator + newURL + File.separator + "pageRank.txt"));
		}
		return -1;
	}

	public String getTitle(String URL){
		String newURL = URL.replace('/','{').replace(':','}').replace('.','(');
		File file = new File("pages" + File.separator + newURL);
		if (file.exists()){
			return readFile("pages" + File.separator + newURL + File.separator + "title.txt");
		}
		return null;
	}

	public double getIDF(String word) {
		String path = "IDFs" + File.separator + word + ".txt";
		File file = new File(path);
		if (file.exists()){
			return Double.parseDouble(readFile(path));
		}
		return 0;
	}

	public double getTF(String URL, String word){
		String newURL = URL.replace('/','{').replace(':','}').replace('.','(');
		File file = new File("pages" + File.separator + newURL);
		if (file.exists()){
			file = new File("pages" + File.separator + newURL + File.separator + "countAll" + File.separator + word + ".txt");
			if (file.exists()){
				return Double.parseDouble(readFile("pages" + File.separator + newURL + File.separator + "countAll" + File.separator + word + ".txt").split(" ")[2]);
			}
		}
		return 0;
	}

	public double getTFIDF(String URL, String word){
		String newURL = URL.replace('/','{').replace(':','}').replace('.','(');
		File file = new File("pages" + File.separator + newURL);
		if (file.exists()){
			file = new File("pages" + File.separator + newURL + File.separator + "countAll" + File.separator + word + ".txt");
			if (file.exists()){
				return Double.parseDouble(readFile("pages" + File.separator + newURL + File.separator + "countAll" + File.separator + word + ".txt").split(" ")[1]);
			}
		}
		return 0;
	}

	public ArrayList<SearchResult> search(String query, boolean boost, int X) {
		searchObj = new Search();
		return searchObj.search(query, boost,X);
	}

//	public List<SearchResult> getSearchResults(){
//		return searchResults;
//
//		Search search = new Search();
//		return search.search(query, boost,X);
//	}

	public static String readFile(String path){
		String result = "";
		try {
			FileInputStream in = new FileInputStream(path);
			while(in.available() > 0){
				result+= (char)in.read() + "";
			}
			in.close();
		} catch (FileNotFoundException e) {
			System.out.println("Error: Cannot open file for reading");
		} catch (IOException e) {
			System.out.println("Error: Cannot read from file");
		}
		return result;
	}

	public static void main(String[] args) {

	}

}
